core/trino-main/src/main/java/io/trino/server/security/HeaderBasedAuthenticator.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.security;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import io.trino.spi.security.AccessDeniedException;
+import io.trino.spi.security.Identity;
+
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import java.security.Principal;
+import java.util.List;
+import java.util.Locale;
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.server.security.UserMapping.createUserMapping;
+import static java.util.Objects.requireNonNull;
+
+public class HeaderBasedAuthenticator
+        implements Authenticator
+{
+    private final HeaderBasedAuthenticatorManager authenticatorManager;
+    private final List<String> headersToExtract;
+    private final boolean treatHeadersCaseSensitive;
```

Comment:
Why do we need this?

Reply:
It was initially added to let the user control if the headers that are being extracted from the request context are matched as-is or converted to their lower case and then matched with the lowercased provided config headers. However, seeing that HTTP 1 [headers are case-insensitive](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) and [upper case headers are malformed](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2) in HTTP 2, it doesn't seem like this would be very useful and so I'll remove it. So, the default behavior would be to lower-case the request context header and match them to the lower-cased config provided headers.

Reply:
> This is only a framework, can we have actual implementation of header based authenticator that could be used out of the box in Trino?

Yes, it is intentional. There isn't a clear-cut standard for authenticating a client based on arbitrary headers like there is for password (basic auth) or certificate (x.509). We are looking to implement this SPI for authenticating on XFCC soon after (might take a few iterations), we'd love to contribute those additions too.